lib/ansible/module_utils/api.py

```diff
@@ -114,3 +115,69 @@ def retried(*args, **kwargs):
 
         return retried
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay_base=3, delay_threshold=60):
+    """The "Full Jitter" backoff strategy.
+
+    Ref: https://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    :param retries: The number of delays to generate.
+    :param delay_base: The base time in seconds used to calculate the exponential backoff.
+    :param delay_threshold: The maximum time in seconds for any delay.
+    """
+    for retry in range(0, retries):
+        yield random.randint(0, min(delay_threshold, delay_base * 2 ** retry))
+
+
+def retry_never(exception_or_result):
+    return False
+
+
+def do_nothing(delay, exception=None, result=None):
+    return
+
+
+def retry_with_delays_and_condition(backoff_iterator, do_on_retry=None, should_retry_error=None, should_retry_result=None):
+    """Generic retry decorator.
+
+    :param backoff_iterator: An iterable of delays in seconds.
+    :param do_on_retry: A callable that takes the current delay, the exception/None, the result/None, and executes before the sleep (returns None).
+    :param should_retry_error: A callable that takes an exception of the decorated function and decides whether to retry or not (returns a bool).
+    :param should_retry_result: A callable that accepts the decorated function return value and decides whether or not to retry (returns a bool).
+    """
+    if do_on_retry is None:
+        do_on_retry = do_nothing
+    if should_retry_error is None:
+        should_retry_error = retry_never
+    if should_retry_result is None:
+        should_retry_result = retry_never
+
+    def function_wrapper(function):
+        def run_function(*args, **kwargs):
+            """This assumes the function has not already been called.
+            If backoff_iterator is empty, we should still run the function a single time with no delay.
+            """
+
+            for delay in backoff_iterator:
+                error = None
+                result = None
+
+                try:
+                    result = function(*args, **kwargs)
+                except Exception as e:
+                    if not should_retry_error(e):
+                        raise
+                    error = e
+                else:
+                    if not should_retry_result(result):
+                        return result
+
+                do_on_retry(delay=delay, exception=error, result=result)
```

Comment:
Looking at this once again, it looks like `do_on_retry()` got too many args and scope creep along with that.

Basically, `should_retry_error()` and `should_retry_result()` can have any side-effects you want already. Do we _really_ need an extra way to do this?

One could do 
```python
def do_common_thing():
    time.sleep(5)

def should_retry_error(error):
    do_common_thing()
    return False

def should_retry_result(result):
    do_common_thing()
    return False

@retry_with_delays_and_condition(
    [...],
    should_retry_error=should_retry_error,
    should_retry_result=should_retry_result
)
```

Reply:
That doesn't accomplish the same thing. Those functions would need to accept the delay. Even if those functions can technically do other stuff, I'd consider it bad code organization because it would be a side-effect in those functions.

The reason `do_on_retry` exists is because I was trying to handle the `reboot` refactor in the future, and it uses the current delay to make helpful debug information during retry attempts. I decided to include `result` because I can see a scenario when you might want to make a `waiter` that also includes debug information. A significant part of this PR could be considered 'scope creep' because it was intended to solve an individual thing and turned into a generic utility, most of which will not be utilized in the backport, for which this is intended.

Reply:
I see. It still feels a bit foreign but I guess we can't do anything about it. Maybe an extra person review would be useful.

Reply:
I decided to strip out everything that has no reason to be backported. With commit history, it will be easy to add those kwargs back when the other places that need those changes are refactored.